### ISSUE TITLE
Restore lost commits to ethereum-master

### ISSUE THREAD

**Issue by placer14** [OpenBazaar/openbazaar-go#1971]
tl;dr: The `ethereum-master` feature branch was force-pushed to an earlier commit on the same tree. The mistake was immediately reported and a set of recommended fixes were executed to repair the branch. It was discovered ~1 month later that the fix was either incomplete or applied improperly. This issue captures the details needed to reconstruct the events which occurred for future triage, support and historical purposes.

On December 9, 2019, it was reported that a contributor `force-pushed the ethereum-master branch from c110900 to fdf79f5`. Based on the reflog shared by that contributor, it is suspected that `ethereum-master` was rebased locally to the earlier commit by accident but not reverted. Then, when force-pushing an unrelated branch using the unspecified form `git push -f` caused all unresolved branches to be overwritten.

Here is a copy of the reflog from that contributor at the time of the force-push. The suspected place where `HEAD` was moved from `c110900` to `fdf79f5` was in the rebase step completing in `HEAD@{21}`.

```
fdf79f529 (HEAD -> ethereum-master, origin/ethereum-master) HEAD@{0}: checkout: moving from compatibilitytest to ethereum-master
6556f74f2 (origin/compatibilitytest, compatibilitytest) HEAD@{1}: checkout: moving from brian.qarunner to compatibilitytest
3e1015d31 (origin/brian.qarunner, brian.qarunner) HEAD@{2}: reset: moving to HEAD
3e1015d31 (origin/brian.qarunner, brian.qarunner) HEAD@{3}: commit: Add QA docker commands to Makefile
dfffbeb72 HEAD@{4}: rebase -i (finish): returning to refs/heads/brian.qarunner
dfffbeb72 HEAD@{5}: rebase -i (squash): QA dockerfile
e3c8a1ed1 HEAD@{6}: rebase -i (pick): QA dockerfile
4f6bac6c5 HEAD@{7}: rebase -i (squash): Add qa.sh script
25300d1e3 HEAD@{8}: rebase -i (squash): # This is a combination of 2 commits.
b97d3425f HEAD@{9}: rebase -i (start): checkout HEAD~5
8d93a0819 HEAD@{10}: commit: Exit bash script if failure
0a3cc22fa HEAD@{11}: reset: moving to 0a3cc22fa8dae666c85a550c4e858cd92b9f9fef
34c0418b7 HEAD@{12}: commit: Exit if test fails
0a3cc22fa HEAD@{13}: checkout: moving from brian.github-actions to brian.qarunner
c9a56a476 (origin/brian.github-actions, brian.github-actions) HEAD@{14}: reset: moving to HEAD
c9a56a476 (origin/brian.github-actions, brian.github-actions) HEAD@{15}: checkout: moving from brian.masterTransports to brian.github-actions
b871985a1 (origin/brian.masterTransports, brian.masterTransports) HEAD@{16}: commit: Fix transports for message retriever and images
b1fb778d8 (origin/master, origin/HEAD, brian.master) HEAD@{17}: checkout: moving from brian.fix-transports to brian.masterTransports
e432a8a6a (origin/brian.fix-transports, brian.fix-transports) HEAD@{18}: commit: Convert custom transports to default transports
c110900e1 HEAD@{19}: rebase finished: refs/heads/brian.fix-transports onto c110900e13ec8d621707e9e12d7c03a73da86018
c110900e1 HEAD@{20}: rebase: checkout origin/ethereum-master
fdf79f529 (HEAD -> ethereum-master, origin/ethereum-master) HEAD@{21}: checkout: moving from brian.hanging-orders to brian.fix-transports
82a99ec54 (origin/brian.hanging-orders, brian.hanging-orders) HEAD@{22}: reset: moving to HEAD
82a99ec54 (origin/brian.hanging-orders, brian.hanging-orders) HEAD@{23}: commit: Update retriever
9e831f6a8 HEAD@{24}: commit: Fix
dae5ebc25 HEAD@{25}: commit: Remove tor transport
720cb9750 HEAD@{26}: commit: Non custom transport
56a109a21 HEAD@{27}: commit: Adjust time
d5b6c6195 HEAD@{28}: commit: Add sleep
064593fb3 HEAD@{29}: commit: MakeFatal
b2ce36f4c HEAD@{30}: commit: Mutex logging
477d73bda (origin/refund) HEAD@{31}: reset: moving to 477d73bdadad86d538762abe5a1c8852030f9c6e
d17135bc9 HEAD@{32}: reset: moving to HEAD
d17135bc9 HEAD@{33}: commit: Add mutex logging
477d73bda (origin/refund) HEAD@{34}: reset: moving to 477d73bdadad86d538762abe5a1c8852030f9c6e
```

[A fix was prescribed](https://ob-1.slack.com/archives/GBHNY4UAH/p1575905572141300), in which the contributor would force-push the branch back to the old commit. The fix looked like:

```
git checkout ethereum-master
git reset --hard <oldsha>
git push origin ethereum-master -f
```

When the fix was prescribed, it was shared with the expectation of teaching the channel about how to recover from this. The contributor applied the fix themselves. The only verification that the fix had be applied properly was asking for the console output from the contributor and verifying the `ethereum-master` branch was pointing to the correct local SHA.

The contributor confirmed the fix was applied like so:
```
$ git push origin ethereum-master -f
Total 0 (delta 0), reused 0 (delta 0)
To https://github.com/OpenBazaar/openbazaar-go
   fdf79f529..c110900e1  ethereum-master -> ethereum-master
```

With the SHAs reversed from their original force-push (originally declared as `force-pushed the ethereum-master branch from c110900 to fdf79f5`) and confirming that branch matched the correct SHA on my local repo, we proceeded with business as usual.

On January 7th, 2020, we discovered some unexpected commits in a long-lived pull request. Some delta was noticed which should have already been merged into the PR's target branch but instead were showing as new additions.

At this point, we began to look more carefully at the fix which was applied and noticed there were other commits that were found to be missing. The following describes the methodology used to construct a pull request.

1. For each SHA before and after the force-push, produce a patch of the SHAs diff against ethereum-master's HEAD. The SHAs of interest are `fdf79f5`, `c110900`, and `ae0282f`. Ex: `git diff -p <target_sha> > patch.file`
2. Apply patches as a 3-way-merge onto ethereum-master and manually resolve conflicts. Ex: `git apply -3 patch.file` (Note: In at least one case, this didn't work as git already detected the SHA had been merged and was applying cleanly as an empty diff. This required a separate working tree checked out at the target SHA to be recursively copied onto a separate `ethereum-master` working tree.)
3. Some changes were found in `vendor`. Due to upstream repos not being properly updated, all OpenBazaar-managed dependencies were checked for changes in each SHA, applied to their upstream repos, and updated. Checked OpenBazaar-managed deps include: `wallet-interface` (updated), `spvwallet` (updated), `go-ethwallet` (updated), `multiwallet` (updated), `bitcoind-wallet` (removed), `zcashd-wallet` (removed).

The outcome of the above steps resulted in commits from `361c5e4..c05ef7e` being applied to `ethereum-master`. In order to prevent future unintended force-pushes against widely-shared branches, we will add protection against force-pushes to `ethereum-master` and be sure to consider the protection on future feature branches.

**Comment by placer14** [comment]
Closed by #1966
